drivers/network/tcpip/tcpip/icmp.c

```diff
@@ -0,0 +1,389 @@
+        RtlZeroMemory(ReplyBuffer, sizeof(*ReplyBuffer));
+        ReplyBuffer->Status = IP_REQ_TIMED_OUT;
+
+        Irp->IoStatus.Status = STATUS_TIMEOUT;
+        Irp->IoStatus.Information = sizeof(*ReplyBuffer);
+    }
+
+    KeReleaseSpinLockFromDpcLevel(&Context->ProcessingLock);
+
+    ExFreePoolWithTag(Context, OUT_DATA_TAG);
+
+    TI_DbgPrint(DEBUG_ICMP, ("Leaving, nReplies: %u\n", ((PICMP_ECHO_REPLY)Irp->AssociatedIrp.SystemBuffer)->Reserved));
+
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+}
+
+NTSTATUS
+DispEchoRequest(
+    _In_ PIRP Irp,
+    _In_ PIO_STACK_LOCATION IrpSp)
+{
+    PICMP_ECHO_REQUEST Request = (PICMP_ECHO_REQUEST)Irp->AssociatedIrp.SystemBuffer;
+    UINT32 ReplySize = IrpSp->Parameters.DeviceIoControl.OutputBufferLength;
+    NTSTATUS Status;
+    TDI_CONNECTION_INFORMATION ConnectionInfo;
+    TA_IP_ADDRESS RemoteAddressTa, LocalAddressTa;
+    PADDRESS_FILE AddrFile;
+    ULONG DataUsed;
+    PUCHAR Buffer;
+    UINT32 RequestSize;
+    PICMP_PACKET_CONTEXT SendContext;
+    KIRQL OldIrql;
+    LARGE_INTEGER TimerResolutionSec;
+
+    TI_DbgPrint(DEBUG_ICMP, ("About to send datagram, ReplySize: %u, SystemBuffer: %p\n", Irp->AssociatedIrp.SystemBuffer));
+
+    if (ReplySize < sizeof(ICMP_ECHO_REPLY))
+    {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    SendContext = ExAllocatePoolWithTag(NonPagedPool, sizeof(*SendContext), OUT_DATA_TAG);
+    if (!SendContext)
+    {
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    RtlZeroMemory(&SendContext->TdiRequest, sizeof(SendContext->TdiRequest));
+    SendContext->TdiRequest.RequestContext = Irp;
+
+    // setting up everything needed for sending the packet
+
+    RtlZeroMemory(&RemoteAddressTa, sizeof(RemoteAddressTa));
+    RtlZeroMemory(&LocalAddressTa, sizeof(LocalAddressTa));
+    RtlZeroMemory(&ConnectionInfo, sizeof(ConnectionInfo));
+
+    RemoteAddressTa.TAAddressCount = 1;
+    RemoteAddressTa.Address[0].AddressLength = TDI_ADDRESS_LENGTH_IP;
+    RemoteAddressTa.Address[0].AddressType = TDI_ADDRESS_TYPE_IP;
+    RemoteAddressTa.Address[0].Address[0].in_addr = Request->Address;
+
+    LocalAddressTa.TAAddressCount = 1;
+    LocalAddressTa.Address[0].AddressLength = TDI_ADDRESS_LENGTH_IP;
+    LocalAddressTa.Address[0].AddressType = TDI_ADDRESS_TYPE_IP;
+    LocalAddressTa.Address[0].Address[0].in_addr = 0;
+
+    Status = FileOpenAddress(&SendContext->TdiRequest, &LocalAddressTa, IPPROTO_ICMP, FALSE, NULL);
+
+    if (!NT_SUCCESS(Status))
+    {
+        TI_DbgPrint(DEBUG_ICMP, ("Failed to open address file status: 0x%x\n", Status));
+
+        ExFreePoolWithTag(SendContext, OUT_DATA_TAG);
+
+        return Status;
+    }
+
+    KeQueryPerformanceCounter(&TimerResolutionSec);
+
+    AddrFile = (PADDRESS_FILE)SendContext->TdiRequest.Handle.AddressHandle;
+
+    // setting up the context
+
+    SendContext->Irp = Irp;
+    SendContext->TimerResolution = TimerResolutionSec.QuadPart / 1000; // now in milliseconds
```

Comment:
Are there issues with the integer overflow here that make this necessary? You'd get a more accurate result if you did this `/ 1000` when calculating `RoundTripTime` in L109 instead (where it would be a `*` rather than a `/`)